src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,493 @@
+variables (𝒜 : ι → submodule R A)
+variable (I : ideal A)
+
+lemma ideal.is_homogeneous.exists_subset_iff_eq_span :
+  (∃ s : set A, s ⊆ set_of (is_homogeneous 𝒜) ∧ I = ideal.span s) ↔
+    I = ideal.span (I ∩ set_of (is_homogeneous 𝒜)) :=
+begin
+  split,
+  { rintros ⟨s, hs, rfl⟩,
+    apply le_antisymm,
+    { exact ideal.span_mono (subset_inter ideal.subset_span hs) },
+    { exact ideal.span_le.2 (inter_subset_left _ _) } },
+  { intros hI,
+    exact ⟨_, inter_subset_right _ _, hI⟩, }
+end
+
+variables [decidable_eq ι] [add_comm_monoid ι]  [graded_algebra 𝒜]
+
+/--An `I : ideal R` is homogeneous if for every `r ∈ I`, all homogeneous components
+  of `r` are in `I`.-/
+def ideal.is_homogeneous : Prop :=
+∀ (i : ι) ⦃r : A⦄, r ∈ I → (graded_algebra.decompose 𝒜 r i : A) ∈ I
+
+lemma ideal.is_homogeneous_iff_forall_subset :
+  ideal.is_homogeneous 𝒜 I ↔ ∀ i, (I : set A) ⊆ graded_algebra.proj 𝒜 i ⁻¹' I :=
+iff.rfl
+
+lemma ideal.is_homogeneous_iff_subset_Inter :
+  ideal.is_homogeneous 𝒜 I ↔ (I : set A) ⊆ ⋂ i, graded_algebra.proj 𝒜 i ⁻¹' ↑I :=
+subset_Inter_iff.symm
+
+lemma ideal.is_homogeneous.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S)) ↔
+  I = ideal.span {x | x ∈ I ∧ is_homogeneous 𝒜 x} :=
+(subtype.exists_set_subtype _).trans (ideal.is_homogeneous.exists_subset_iff_eq_span _ _)
+
+lemma mul_homogeneous_element_mem_of_mem
+  {I : ideal A} (r x : A) (hx₁ : is_homogeneous 𝒜 x) (hx₂ : x ∈ I) (j : ι) :
+  graded_algebra.proj 𝒜 j (r * x) ∈ I :=
+begin
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw [←graded_algebra.sum_support_decompose 𝒜 r, finset.sum_mul, linear_map.map_sum],
+  apply ideal.sum_mem,
+  intros k hk,
+  obtain ⟨i, hi⟩ := hx₁,
+  have mem₁ : (graded_algebra.decompose 𝒜 r k : A) * x ∈ 𝒜 (k + i) := graded_monoid.mul_mem
+    (submodule.coe_mem _) hi,
+  rw [graded_algebra.proj_apply, graded_algebra.decompose_of_mem 𝒜 mem₁,
+    coe_of_submodule_apply 𝒜, submodule.coe_mk],
+  split_ifs,
+  { exact I.mul_mem_left _ hx₂ },
+  { exact I.zero_mem },
+end
+
+lemma ideal.is_homogeneous.iff_eq :
+  ideal.is_homogeneous 𝒜 I ↔ I = ideal.span {x | x ∈ I ∧ is_homogeneous 𝒜 x} :=
+⟨ λ hI, begin
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  ext, split; intro hx,
+  { rw ←graded_algebra.sum_support_decompose 𝒜 x,
+    refine ideal.sum_mem _ _,
+    intros j hj,
+    rw ideal.mem_span, intros J HJ,
+    refine HJ ⟨hI j hx, j, submodule.coe_mem _⟩, },
+  { rw [ideal.mem_span] at hx,
+    apply hx,
+    exact inter_subset_left _ _, },
+  end,
+  λ hI, begin
+    intros i r hr,
+    rw ←graded_algebra.proj_apply,
+    rw [ideal.span, finsupp.span_eq_range_total] at hI,
+    rw hI at hr,
+    obtain ⟨s, rfl⟩ := hr,
+    simp_rw [finsupp.total_apply, finsupp.sum, linear_map.map_sum, smul_eq_mul],
+    refine ideal.sum_mem I _,
+    rintros ⟨j, ⟨hj₁, hj₂⟩⟩ hj₃,
+    exact mul_homogeneous_element_mem_of_mem _ _ _ hj₂ hj₁ _,
+  end ⟩
```

Comment:
This reverse direction should become `rw hI, exact ideal.is_homogeneous.homogeneous_core`, where `ideal.is_homogeneous.homogeneous_core` is proved before this lemma instead of vice versa.

Reply:
done